File: two_sigma/financial_modeling_challenge/cpp/tests/test_solution.cpp
#include <gtest/gtest.h>
#include "model.h"
#include "environment.h"
#include "data_generator.h"
#include <filesystem>

using namespace two_sigma;

TEST(FinancialModelTest, TrainAndPredict) {
    FinancialModel model(0.1);
    Eigen::MatrixXd X(10, 5);
    X.setRandom();
    Eigen::VectorXd y(10);
    y.setRandom();
    
    model.train(X, y);
    Eigen::VectorXd predictions = model.predict(X);
    EXPECT_EQ(predictions.size(), 10);
}

TEST(DataGeneratorTest, GenerateFile) {
    std::string path = "test_data.csv";
    DataGenerator gen(100, 10, 5);
    gen.generate(path);
    EXPECT_TRUE(std::filesystem::exists(path));
    std::filesystem::remove(path);
}

TEST(EnvironmentTest, Lifecycle) {
    std::string path = "env_test.csv";
    DataGenerator gen(200, 10, 5);
    gen.generate(path);
    
    Environment env(path);
    auto obs = env.reset();
    EXPECT_GT(obs.train_features.rows(), 0);
    EXPECT_GT(obs.test_features.rows(), 0);
    
    Eigen::VectorXd preds(obs.test_features.rows());
    preds.setConstant(0.01);
    
    auto [next_obs, reward] = env.step(preds);
    EXPECT_FALSE(env.is_done());
    
    std::filesystem::remove(path);
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: two_sigma/financial_modeling_challenge/rust/tests/test_solution.rs
use two_sigma_financial_modeling::model::FinancialModel;
use two_sigma_financial_modeling::data_generator::DataGenerator;
use two_sigma_financial_modeling::environment::Environment;
use ndarray::{Array1, Array2};
use std::fs;

#[test]
fn test_model_train_predict() {
    let mut model = FinancialModel::new(0.1);
    let x = Array2::from_elem((10, 5), 1.0);
    let y = Array1::from_elem(10, 2.0);
    model.train(&x, &y);
    let preds = model.predict(&x);
    assert_eq!(preds.len(), 10);
}

#[test]
fn test_data_generation() {
    let path = "test_gen.csv";
    let gen = DataGenerator::new(100, 10, 5);
    gen.generate(path).unwrap();
    assert!(fs::metadata(path).is_ok());
    fs::remove_file(path).unwrap();
}

#[test]
fn test_environment_flow() {
    let path = "test_env.csv";
    let gen = DataGenerator::new(200, 10, 5);
    gen.generate(path).unwrap();
    
    let mut env = Environment::new(path).unwrap();
    let obs = env.reset();
    assert!(obs.train_features.nrows() > 0);
    
    let preds = Array1::from_elem(obs.test_features.nrows(), 0.01);
    let (next, _, _) = env.step(preds);
    assert!(next.is_some());
    
    fs::remove_file(path).unwrap();
}
